### PR TITLE
feat: Add a blst-based Groth16 verifier

### PR DESCRIPTION
**Pull request by huitseeker**
This PR sits on top of #110 which should be merged before it.
Adds a pure-blst Groth16 verifier. Approx verification time ~1/2 that of Arkworks.

Detailed bench output at https://gist.github.com/de4e15a3c20cb5251dbe1fc8ecef190d

### CODE REVIEW

File: fastcrypto-zkp/src/verifier.rs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    let mut pairing_blst = Pairing::new(false, &dst);
+    pairing_blst.raw_aggregate(&blst_proof_b, &blst_proof_a);
+    pairing_blst.raw_aggregate(&blst_proof_1_g2, &blst_proof_1_g1);
+    pairing_blst.raw_aggregate(&blst_proof_2_g2, &blst_proof_2_g1);
+    pairing_blst.as_fp12().final_exp()
+}
+
+/// Returns the validity of the Groth16 proof passed as argument. The format of the inputs is assumed to be in arkworks format.
+/// See [`multipairing_with_processed_vk`] for the actual pairing computation details.
+/// TODO: due to arkworks incompatibilities in BLS12-381 point (de) serialization, we should probably implement a custom (de)serialization
```
Comment (kchalkias) [inline fastcrypto-zkp/src/verifier.rs:225]: <3

**Review by kchalkias — COMMENTED** [review]
Ooc, do conversions introduce any delays and ideally a ZK verifier framework working directly with BLST's element format would result to even better bench numbers?

**Comment by huitseeker** [comment]
@kchalkias to answer your question about the cost of conversions precisely, I have added benchmarks for the conversion operations (both ways, Arkworks <-> BLST).

The on-chain verification converts as many scalars (in code `BlsFr -> blst_fr`) as there are public inputs, along with the three points (in code, 2x `G1Affine -> blst_p1_affine`, 1x `G2_affine -> blst_p2_affine`), and returns one result 

Here is a sample of the conversion benchmark:
```
Conversions Arkworks -> Blst/BlsFr -> blst_fr
                        time:   [29.572 ns 29.608 ns 29.645 ns]
Conversions Arkworks -> Blst/G1Affine -> blst_p1_affine
                        time:   [426.95 ns 427.46 ns 428.02 ns]
Conversions Arkworks -> Blst/G2Affine -> blst_p2_affine
                        time:   [974.00 ns 975.46 ns 976.98 ns]
```

To give an order of magnitude, our benchmarks for both the preparation of the verification key and the proof verification are in the hundreds of microseconds range. 

Given this, I do not think the cost of conversions is relevant.

**Review by kchalkias — APPROVED** [review]

File: fastcrypto-zkp/benches/conversions.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    grp.bench_with_input("Fq2 -> blst_fp2", &element, |b, element| {
+        b.iter(|| bls_fq2_to_blst_fp2(element));
+    });
+
+    let element = Fq6::rand(&mut rng);
+    grp.bench_with_input("Fq6 -> blst_fp6", &element, |b, element| {
+        b.iter(|| bls_fq6_to_blst_fp6(element));
+    });
+
+    let element = Fq12::rand(rng);
```
[the remark this answers is not in the document]
Comment (huitseeker) [reply fastcrypto-zkp/benches/conversions.rs:40]: It does. I just defined `rng` quickly as an `&mut`.

File: fastcrypto-zkp/src/unit_tests/verifier_tests.rs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    );
+
+    Bls12_381::final_exponentiation(&qap).unwrap()
+}
+
+const LEN: usize = 10;
+
+proptest! {
+    // This technical test is necessary because blst does not expose
+    // the miller_loop_n operation, and forces us to abuse the signature-oriented pairing engine
```
[the remark this answers is not in the document]
Comment (huitseeker) [reply fastcrypto-zkp/src/unit_tests/verifier_tests.rs:82]: https://github.com/supranational/blst/issues/136

File: fastcrypto-zkp/src/unit_tests/verifier_tests.rs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    fn test_g1_linear_combination(
+        frs in collection::vec(arb_bls_fr(), LEN-1..=LEN-1),
+        a_s in collection::vec(arb_bls_g1_affine(), LEN..=LEN),
+    ) {
+
+        let pts: Vec<blst_p1_affine> = a_s
+            .iter()
+            .map(bls_g1_affine_to_blst_g1_affine)
+            .collect();
+        // TODO: find a more direct way to grab this constant
```
[the remark this answers is not in the document]
Comment (huitseeker) [reply fastcrypto-zkp/src/unit_tests/verifier_tests.rs]: Fix'd!

**Pull request merged**